pkg/agent/proxy/proxier.go

```diff
@@ -548,41 +643,27 @@ func (p *proxier) installServices() {
 		}
 
 		withSessionAffinity := svcInfo.SessionAffinityType() == corev1.ServiceAffinityClientIP
-		internalPolicyLocal := svcInfo.InternalPolicyLocal()
-		externalPolicyLocal := svcInfo.ExternalPolicyLocal()
-		var internalGroupID, externalGroupID, clusterGroupID binding.GroupIDType
-		// Ensure a group for internal traffic exist.
-		if internalGroupID, ok = p.installServiceGroup(svcPortName, needUpdateEndpoints, internalPolicyLocal, withSessionAffinity, localEndpoints, clusterEndpoints); !ok {
-			continue
+		var localGroupID, clusterGroupID binding.GroupIDType
+		// categorizeEndpoints has checked if localGroup and clusterGroup should exist. We just create the group if its
+		// Endpoints is not nil.
+		// Note that nil represents the group should not exist and empty represents the group should exist but there is
+		// no available Endpoints.
+		if localEndpoints != nil {
+			if localGroupID, ok = p.installServiceGroup(svcPortName, needUpdateEndpoints, true, withSessionAffinity, false, localEndpoints); !ok {
+				continue
+			}
+		} else {
+			if !p.removeServiceGroup(svcPortName, true) {
+				continue
+			}
 		}
-		// Ensure a group for external traffic exist if it's externally accessible, and remove the unneeded group.
-		if svcInfo.ExternallyAccessible() {
-			if externalPolicyLocal != internalPolicyLocal {
-				if externalGroupID, ok = p.installServiceGroup(svcPortName, needUpdateEndpoints, externalPolicyLocal, withSessionAffinity, localEndpoints, clusterEndpoints); !ok {
-					continue
-				}
-				if externalPolicyLocal {
-					clusterGroupID = internalGroupID
-				} else {
-					clusterGroupID = externalGroupID
-				}
-			} else {
-				externalGroupID = internalGroupID
-				if externalPolicyLocal {
-					if clusterGroupID, ok = p.installServiceGroup(svcPortName, needUpdateEndpoints, false, withSessionAffinity, nil, clusterEndpoints); !ok {
-						continue
-					}
-				} else {
-					// Ensure the other group is removed as ExternalTrafficPolicy is the same as InternalTrafficPolicy.
-					if !p.removeServiceGroup(svcPortName, !internalPolicyLocal) {
-						continue
-					}
-					clusterGroupID = externalGroupID
-				}
+		if clusterEndpoints != nil {
+			// Only DSR mode cares the selection method, we use "hash" method as long as DSR is enabled.
```

Comment:
Shouldn't we be more conservative and only use `hash` for DSR Services? I remember that when you did the performance testing and evaluating several scenarios, `dp_hash` was superior to `hash` **in the absence of a learn action** (but at the time you were not masking the source port for hash). In particular, the latency was higher?